### PR TITLE
Breeding system implementation

### PR DESCRIPTION
**Pull request by shreyasiyer14**
- [x] Implement gooey follow mechanic
- [x] Create Love Shack ST using `Valentines` module
- [x] Implement breeding recipes
- [x] Add egg dropping/hatching mechanic

### CODE REVIEW

**Comment by mjuvekar7** [comment]
In my opinion, the follow mechanism should be an independent mechanic in itself, and not exclusively for mating. I can see a bunch of other potential uses for it

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 1]
```diff
@@ -96,13 +96,17 @@ public void onBreedingGooey(BreedGooeyEvent event, EntityRef gooeyEntity, GooeyC
...
             if (!itemName.isEmpty() && hungerComponent.food.contains(itemName)) {
-                logger.info("Selected for breeding ...");
+                logger.info("Selected for breeding...");
                 FollowComponent followComponent = gooeyEntity.getComponent(FollowComponent.class);
                 BehaviorComponent behaviorComponent = gooeyEntity.getComponent(BehaviorComponent.class);
                 MatingComponent matingComponent = new MatingComponent();
 
                 followComponent.entityToFollow = event.getInstigator();
-                behaviorComponent.tree = assetManager.getAsset("GooKeeper:breedingBehavior", BehaviorTree.class).get();
+                if (!matingComponent.selectedForMating) {
```
Comment (mjuvekar7) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: You could remove lines 105 to 109 and replace them with just `matingComponent.selectedForMating = !matingComponent.selectedForMating`

File: assets/ui/gooeyActivateScreen.ui [hunk 2]
```diff
@@ -17,8 +17,8 @@
           },
           {
             "type": "UIButton",
-            "text": "Select gooey for breeding",
-            "id": "breedButton"
+            "text": "Lure the gooey to follow",
```
Comment (mjuvekar7) [inline assets/ui/gooeyActivateScreen.ui]: This is too verbose for button test. Just "Lure" would look much cleaner, and it's descriptive enough so the player won't get confused.

File: src/main/java/org/terasology/gookeeper/component/BreedingBlockComponent.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.gookeeper.component;
+
+import org.terasology.entitySystem.Component;
+import org.terasology.entitySystem.entity.EntityRef;
+
+public class BreedingBlockComponent implements Component {
```
Comment (mjuvekar7) [inline src/main/java/org/terasology/gookeeper/component/BreedingBlockComponent.java:21]: Javadoc the class declaration as well

File: src/main/java/org/terasology/gookeeper/component/BreedingBlockComponent.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    /**
+     * The gooey entity which is currently standing on the breeding block
```
Comment (mjuvekar7) [inline src/main/java/org/terasology/gookeeper/component/BreedingBlockComponent.java]: Too verbose. This could instead be just "The gooey entity which is to be involved in mating."

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 5]
```diff
@@ -77,9 +60,41 @@
...
     private NUIManager nuiManager;
 
+    @In
+    private EntityManager entityManager;
+
     private static final Logger logger = LoggerFactory.getLogger(BreedingSystem.class);
     private Random random = new FastRandom();
 
+    /**
+     * This method is called upon when a gooey that is following the player is brought to
```
Comment (mjuvekar7) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: This whole description could be just "This method is called when a gooey steps on a breeding block." Breeding blocks are found only inside love shacks, so you don't need to specify that. Similarly, gooeys won't go into the love shack unless they are brought there, so you don't need to specify that either.

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 6]
```diff
@@ -77,9 +60,41 @@
...
+     * the love shack and it steps over a breeding block.
+     *
+     * Receives OnEnterBlockEvent when a gooey entity steps over a breeding block.
+     *
+     * @param event     The OnEnterBlockEvent
+     * @param entity    the gooey entity
+     */
+    @ReceiveEvent(components = {GooeyComponent.class})
+    public void onEnterBlock(OnEnterBlockEvent event, EntityRef entity) {
+        LocationComponent loc = entity.getComponent(LocationComponent.class);
```
Comment (mjuvekar7) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: Since you're never actually using the variable `loc` you can remove this line and rewrite the next line as
`Vetor3f pos = entity.getComponent(LocationComponent.class).getWorldPosition();`

Your choice. If you feel the current one is more readable that's fine too.

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 7]
```diff
@@ -88,30 +103,23 @@
...
-                gooeyEntity.saveComponent(behaviorComponent);
-                gooeyEntity.addOrSaveComponent(matingComponent);
-                gooeyEntity.saveComponent(characterMovementComponent);
+    public void onBreedingGooey(BreedGooeyEvent event, EntityRef gooeyEntity, GooeyComponent gooeyComponent) {
+        logger.info("Selected for breeding...");
+        MatingComponent matingComponent = new MatingComponent();
+
+        matingComponent.selectedForMating = !matingComponent.selectedForMating;
+
+        for (EntityRef breedingBlock : entityManager.getEntitiesWith(BreedingBlockComponent.class)) {
```
Comment (mjuvekar7) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: This for loop should not get executed if selectedForMating is false? From what I understand, as it is right now, if the player deselcts a gooey for mating, line 110 will toggle `selectedForMating` to false (which is right), but this `for` loop will still get executed (which you do not want) and thus select a partner for mating

Comment (mjuvekar7) [reply src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: Maybe you want to enclose this whole loop inside `if (matingComponent.selectedForMating) {...}`

**Review by mjuvekar7 — CHANGES_REQUESTED** [review]

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java (hunk 6 above)
Comment (syntaxi) [reply src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: Alternatively you could also pop it into the method signature

File: src/main/java/org/terasology/gookeeper/event/FollowGooeyEvent.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public EntityRef getGooey() {
+        return gooey;
+    }
+
+    public EntityRef getInstigator() {
+        return instigator;
+    }
+}
```
Comment (mjuvekar7) [inline src/main/java/org/terasology/gookeeper/event/FollowGooeyEvent.java]: Minor checkstyle issue: add a newline at the end of the file (same for all other files)

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java (hunk 6 above)
Comment (shreyasiyer14) [reply src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: I feel it is more readable, so I'll stick with it for now 😄

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 9]
```diff
@@ -165,11 +172,82 @@ public void onBreedingGooey(SelectForBreedingEvent event, EntityRef gooeyEntity,
...
+        }
+
+        if (parent2Friendly != null) {
+            friendlyComponent.friendlinessFactor += parent2Friendly.friendlinessFactor;
+        }
+
+        friendlyComponent.friendlinessFactor /= 2f;
+
+        Component dominantComponent = getDominantComponent(aggressiveComponent, neutralComponent, friendlyComponent);
+
```
Comment (mjuvekar7) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: No newline

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 10]
```diff
@@ -165,11 +172,82 @@ public void onBreedingGooey(SelectForBreedingEvent event, EntityRef gooeyEntity,
...
+        if (parent1Aggressive != null) {
+            aggressiveComponent.aggressivenessFactor += parent1Aggressive.aggressivenessFactor;
+        }
+
+        if (parent2Aggressive != null) {
+            aggressiveComponent.aggressivenessFactor += parent2Aggressive.aggressivenessFactor;
+        }
+
+        aggressiveComponent.aggressivenessFactor /= 2f;
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: This may be overly complicated, but you've got a decent chunk of repeated code here that could be extracted and simplified.

Consider making an abstract component with a single public field:
```java
public abstract FactorComponent implements Component {
    public float magnitude;
}
```

Then, make `AggressiveComponent`, `NeutralComponent`, and `FriendlyComponent` all extend `FactorComponent`

Next, make a new method that takes a given class adds the two factors together, into a new component. Something like

```java
public <T extends FactorComponent> T getChildFactor(T childComponent, EntityRef parent1, EntityRef parent2) {
    FactorComponent source1 = parent1.getComponent(childComponent.getClass());
    FactorComponent source2 = parent2.getComponent(childComponent.getClass());
    /* Do null checks and add them to `childComponent` */
    /* Divide by scale factor /*
    /* Return child component */
}
```

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 11]
```diff
@@ -165,11 +172,82 @@ public void onBreedingGooey(SelectForBreedingEvent event, EntityRef gooeyEntity,
...
+            friendlyComponent.friendlinessFactor += parent1Friendly.friendlinessFactor;
+        }
+
+        if (parent2Friendly != null) {
+            friendlyComponent.friendlinessFactor += parent2Friendly.friendlinessFactor;
+        }
+
+        friendlyComponent.friendlinessFactor /= 2f;
+
+        Component dominantComponent = getDominantComponent(aggressiveComponent, neutralComponent, friendlyComponent);
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: Additionally to my above comment, you may find it easier to have a single component with a scale from aggression to friendliness.
Something like `AttitudeComponent` with a single float/int value. Use negatives as aggression and positives as friendliness, and zero as neutral (or rather, being within a set distance of zero)
Then you can just add the two parent's values, and you'll get the kids value.

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 12]
```diff
@@ -15,10 +15,12 @@
  */
 package org.terasology.gookeeper.system;
 
+import com.sun.xml.internal.bind.v2.runtime.reflect.opt.Const;
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: Whoa, why is this here?

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 13]
```diff
@@ -28,17 +30,19 @@
 import org.terasology.entitySystem.systems.BaseComponentSystem;
 import org.terasology.entitySystem.systems.RegisterMode;
 import org.terasology.entitySystem.systems.RegisterSystem;
-import org.terasology.gookeeper.component.BreedingBlockComponent;
-import org.terasology.gookeeper.component.GooeyComponent;
-import org.terasology.gookeeper.component.MatingComponent;
+import org.terasology.gookeeper.Constants;
+import org.terasology.gookeeper.component.*;
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: Begone star imports!

**Review by mjuvekar7 — APPROVED** [review]

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 14]
```diff
@@ -328,10 +346,29 @@ private Material getOffspringMaterial(EntityRef parent1, EntityRef parent2) {
...
+
+        offspringEntity.addOrSaveComponent(offspringGooeyColor);
+        offspringMaterial = Assets.getMaterial(materialName).get();
 
-        return offspringMaterial;
+        if (offspringMaterial != null) {
+            return offspringMaterial;
+        } else {
+            return null;
+        }
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: This can be simplified to `Assets.getMaterial(materialName).orElse(null)`

Additionally as per [the `get` javadocs](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#get) the call will throw an error if the value was null. Judging by how you are handling the case where the value is null this is probably something to consider

File: assets/materials/1y.mat [hunk 15]
```diff
@@ -0,0 +1,8 @@
+{
+    "shader" : "engine:genericMeshMaterial",
+    "params" : {
+        "diffuse" : "yellowGooey",
```
Comment (syntaxi) [inline assets/materials/1y.mat]: Seeing as how you've named all the other child textures as `1r2b` and the materials follow the same pattern, I'd recommend renaming the texture to `1y` here (and `1r` & `1b` likewise)

File: src/main/java/org/terasology/gookeeper/actions/AfterBreedingAction.java [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+
+
+@BehaviorAction(name = "AfterBreeding")
+public class AfterBreedingAction extends BaseAction {
+
+    @In
+    private Time time;
+
+    @In
+    private LocalPlayer localPlayer;
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/actions/AfterBreedingAction.java]: I'm assuming that this and `Time` are intended to be used and this class is not fully implemented?
If not then these shouldn't be injected

File: src/main/java/org/terasology/gookeeper/actions/BeginBreedingAction.java [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+
+@BehaviorAction(name = "BeginBreeding")
+public class BeginBreedingAction extends BaseAction {
+
+    @In
+    private Time time;
+
+    @In
+    private LocalPlayer localPlayer;
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/actions/BeginBreedingAction.java]: Similar to above, if these aren't intended to be used they should be removed.

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 18]
```diff
@@ -19,55 +19,69 @@
...
 import org.terasology.utilities.random.FastRandom;
 import org.terasology.utilities.random.Random;
 import org.terasology.world.BlockEntityRegistry;
-import org.terasology.world.WorldProvider;
-import org.terasology.worldlyTooltipAPI.events.GetTooltipIconEvent;
-import org.terasology.worldlyTooltipAPI.events.GetTooltipNameEvent;
+import sun.text.resources.en.FormatData_en_IE;
+
+import java.math.RoundingMode;
+import java.util.Vector;
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: Is this import used?
Terasology uses it's own vectors so I'm unsure as to why this exists

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 19]
```diff
@@ -19,55 +19,69 @@
...
-import org.terasology.rendering.nui.layers.ingame.inventory.GetItemTooltip;
-import org.terasology.rendering.nui.widgets.TooltipLine;
 import org.terasology.utilities.Assets;
 import org.terasology.utilities.random.FastRandom;
 import org.terasology.utilities.random.Random;
 import org.terasology.world.BlockEntityRegistry;
-import org.terasology.world.WorldProvider;
-import org.terasology.worldlyTooltipAPI.events.GetTooltipIconEvent;
-import org.terasology.worldlyTooltipAPI.events.GetTooltipNameEvent;
+import sun.text.resources.en.FormatData_en_IE;
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: Another sneaky import here

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 20]
```diff
@@ -77,41 +91,295 @@
...
+     * This method is called when a gooey steps on a breeding block
+     *
+     * @param event     The OnEnterBlockEvent
+     * @param entity    the gooey entity
+     */
+    @ReceiveEvent
+    public void onEnterBlock(OnEnterBlockEvent event, EntityRef entity, GooeyComponent gooeyComponent) {
+        LocationComponent loc = entity.getComponent(LocationComponent.class);
+        Vector3f pos = loc.getWorldPosition();
+        pos.setY(pos.getY() -1);
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java:116]: There is a helper method for this.
I believe it's `pos.addY(-1)`

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 21]
```diff
@@ -77,41 +91,295 @@
...
-     * Receives the BreedGooeyEvent when the "activated" gooey is chosen for breeding.
+     * This method is called when a gooey steps on a breeding block
+     *
+     * @param event     The OnEnterBlockEvent
+     * @param entity    the gooey entity
+     */
+    @ReceiveEvent
+    public void onEnterBlock(OnEnterBlockEvent event, EntityRef entity, GooeyComponent gooeyComponent) {
+        LocationComponent loc = entity.getComponent(LocationComponent.class);
+        Vector3f pos = loc.getWorldPosition();
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java:115]: It may make it clearer if you convert to `Vector2i` here, rather than in the `BlockEntityManager` call below

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 22]
```diff
@@ -77,41 +91,295 @@
...
+        pos.setY(pos.getY() -1);
+
+        EntityRef blockEntity = blockEntityRegistry.getExistingBlockEntityAt(new Vector3i(pos, RoundingMode.HALF_UP));
+
+        if (blockEntity.hasComponent(BreedingBlockComponent.class)) {
+            FollowComponent followComponent = entity.getComponent(FollowComponent.class);
+            BreedingBlockComponent breedingBlockComponent = blockEntity.getComponent(BreedingBlockComponent.class);
+
+            if (followComponent.entityToFollow != EntityRef.NULL && breedingBlockComponent.parentGooey == EntityRef.NULL && gooeyComponent.isCaptured) {
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java:124]: The null checks can be simplified to `followComponent.entityToFollow.exists()`, likewise so can the breeding block entity check (negating of course)

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 23]
```diff
@@ -77,41 +91,295 @@
...
+
+        if (!gooeyEntity.hasComponent(MatingComponent.class)) {
+            matingComponent = new MatingComponent();
+        } else {
+            matingComponent = gooeyEntity.getComponent(MatingComponent.class);
+        }
+
+        matingComponent.selectedForMating = !matingComponent.selectedForMating;
+
+        if (matingComponent.selectedForMating) {
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java:153]: This is a fairly big `if`.
 It may be worth moving the contents of it to a separate method, named appropriately

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 24]
```diff
@@ -77,41 +91,295 @@
...
+        if (characterMovementComponent != null) {
+            characterMovementComponent.speedMultiplier = 1f;
+        }
+
+        if (characterMovementComponent1 != null) {
+            characterMovementComponent1.speedMultiplier = 1f;
+        }
+
+        gooeyEntity.saveComponent(characterMovementComponent);
+        matingComponent.matingWithEntity.saveComponent(characterMovementComponent1);
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: Shouldn't these lines be inside the prior null checking `if`s?

File: src/main/java/org/terasology/gookeeper/system/BreedingSystem.java [hunk 25]
```diff
@@ -77,41 +91,295 @@
...
+
+    /**
+     * Receives DelayedActionTriggeredEvent, which spawns the gooey egg
+     *
+     * @param event     The DelayedActionTriggeredEvent event
+     * @param gooeyEntity    The entity to which the event is sent
+     */
+    @ReceiveEvent
+    public void onDelayedAction(DelayedActionTriggeredEvent event, EntityRef gooeyEntity) {
+        if (event.getActionId().equals(Constants.spawnGooeyEggEventID)) {
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/system/BreedingSystem.java]: Likewise to above, both this and the `else` section should be moved to separate methods

File: src/main/java/org/terasology/gookeeper/system/GooeySystem.java [hunk 26]
```diff
@@ -456,4 +463,36 @@ public void onBump(HorizontalCollisionEvent event, EntityRef entity) {
...
+     * @param event
+     * @param gooeyEntity
+     * @param gooeyComponent
+     * @param hungerComponent
+     */
+    @ReceiveEvent
+    public void onGooeyFollowPlayer(FollowGooeyEvent event, EntityRef gooeyEntity, GooeyComponent gooeyComponent, HungerComponent hungerComponent) {
+        CharacterHeldItemComponent characterHeldItemComponent = event.getInstigator().getComponent(CharacterHeldItemComponent.class);
+
+        if (characterHeldItemComponent != null && characterHeldItemComponent.selectedItem.getComponent(DisplayNameComponent.class) != null) {
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/system/GooeySystem.java]: Second check should probably be replaced with `hasComponent`

File: src/main/java/org/terasology/gookeeper/system/HungerSystem.java [hunk 27]
```diff
@@ -193,11 +191,15 @@ public void addAttributesToTooltip(GetItemTooltip event, EntityRef entity, Gooey
...
             event.getTooltipLines().add(new TooltipLine("Can eat : " + food));
         }
+        event.getTooltipLines().add(new TooltipLine("PlazMaster frequency required: " + gooeyComponent.stunFrequency));
     }
 
     @ReceiveEvent(components = GooeyComponent.class)
     public void setIcon(GetTooltipIconEvent event, EntityRef entityRef) {
-        event.setIcon(Assets.getTextureRegion("GooKeeper:"+ entityRef.getComponent(DisplayNameComponent.class).name + "Tooltip").get());
+        TextureRegion textureRegion = Assets.getTextureRegion("GooKeeper:"+ entityRef.getComponent(DisplayNameComponent.class).name + "Tooltip").get();
+        if (textureRegion != null) {
```
Comment (syntaxi) [inline src/main/java/org/terasology/gookeeper/system/HungerSystem.java]: Again, use `Optional.isPresent`

**Review by syntaxi — CHANGES_REQUESTED** [review]
Couple of things I'd like touched up.
